daisi/src/cpps/logical/algorithms/disposition/round_robin_initiator.cpp

```diff
@@ -0,0 +1,218 @@
+// the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+// Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along with DAISI. If not, see
+// <https://www.gnu.org/licenses/>.
+//
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include "round_robin_initiator.h"
+
+#include "cpps/amr/model/amr_fleet.h"
+
+namespace daisi::cpps::logical {
+RoundRobinInitiator::RoundRobinInitiator(std::shared_ptr<sola_ns3::SOLAWrapperNs3> sola,
+                                         std::shared_ptr<CppsLoggerNs3> logger)
+    : DispositionInitiator(sola, logger) {
+  sola::Request amr_request;
+  amr_request.all = true;
+  amr_request.permissive = true;
+  amr_request.request = "(servicetype == transport)";
+
+  amr_find_result_ = sola_->findService(amr_request);
+};
+
+bool RoundRobinInitiator::process(const AssignmentResponse &assignment_response) {
+  if (assignment_response.doesAccept()) {
+    assignment_acceptions_.push_back(assignment_response);
+    // TODO find correct task for corresponding uuid
+    // logMaterialFlowOrderStatesOfTask(assignment_response.getTaskUuid(), OrderStates::kQueued);
+  } else {
+    throw std::logic_error("the AMR should always accept the task assignment!");
+  }
+  return true;
+}
+
+// TODO
+bool RoundRobinInitiator::process(const StatusUpdate &status_update) { return true; }
+
+void RoundRobinInitiator::addMaterialFlow(std::shared_ptr<material_flow::MFDLScheduler> scheduler) {
+  if (!preparation_finished_) {
+    if (amr_find_result_.valid()) {
+      readAmrRequestFuture();
+    } else {
+      throw std::runtime_error("Future is not ready yet but required.");
+    }
+  }
+
+  material_flows_.push_back(*scheduler);
+  distributeMFTasks(material_flows_.size() - 1, false);
+}
+
+void RoundRobinInitiator::readAmrRequestFuture() {
+  for (const auto &result : amr_find_result_.get()) {
+    ParticipantInfo info;
+    float max_payload = -1;
+    std::optional<amr::LoadCarrier> loadcarrier;
+
+    for (const auto &entry : result) {
+      std::string key = std::get<0>(entry);
+      auto value = std::get<1>(entry);
+      if (key == "endpoint") {
+        info.connection_string = std::get<std::string>(value);
+      } else if (key == "loadcarriertype") {
+        loadcarrier = amr::LoadCarrier(std::get<std::string>(value));
+      } else if (key == "maxpayload") {
+        max_payload = std::get<float>(value);
+      }
+
+      if (max_payload != -1 && loadcarrier != std::nullopt) {
+        amr::AmrStaticAbility ability = amr::AmrStaticAbility(loadcarrier.value(), max_payload);
+        info.ability = ability;
+      }
+    }
+    if (info.connection_string.empty() ||
+        info.ability.getLoadCarrier().getTypeAsString() == "invalid") {
+      // something is still uninitialized
+      throw std::runtime_error("The AMR's ability and connection string should be known.");
+    }
```

Comment:
Move this kind of check into the ParticipantInfo class and use something like `info.isValid()`. Move as much logic from the code into the smaller classes to make it more modular and easy to read. 

You could do something similar for the `AmrStaticAbility` if you want.